packages/web/src/features/translation/TranslationView.tsx

```diff
@@ -1,42 +1,137 @@
 import apiClient from '../../shared/apiClient';
 import { useLayoutContext } from '../../app/Layout';
-import {
-  GetVerseGlossesResponseBody,
-  GetVerseResponseBody,
-} from '@translation/api-types';
-
-export async function translationViewLoader({ params }: LoaderFunctionArgs) {
-  const verse = await apiClient.verses.findById(params.verseId ?? 'unknown');
-  const referenceGlosses = await apiClient.verses.findVerseGlosses(
-    params.verseId ?? 'unknown',
-    'en'
+import TextInput from '../../shared/components/TextInput';
+import { useState } from 'react';
+import InputHelpText from '../../shared/components/InputHelpText';
+import { GetVerseGlossesResponseBody } from '@translation/api-types';
+import { Icon } from '../../shared/components/Icon';
+
+export default function TranslationView() {
+  const params = useParams() as { verseId: string };
+  const { language } = useLayoutContext();
+
+  const verseQuery = useQuery(['verse', params.verseId], () =>
+    apiClient.verses.findById(params.verseId)
+  );
+  const referenceGlossesQuery = useQuery(
+    ['verse-glosses', 'en', params.verseId],
+    () => apiClient.verses.findVerseGlosses(params.verseId, 'en')
+  );
+  const targetGlossesQuery = useQuery(
+    ['verse-glosses', language, params.verseId],
+    () => apiClient.verses.findVerseGlosses(params.verseId, language)
   );
 
-  return { verse, referenceGlosses };
-}
+  const [glossRequests, setGlossRequests] = useState<
+    { wordId: string; requestId: number }[]
+  >([]);
+  const queryClient = useQueryClient();
+  const glossMutation = useMutation({
+    mutationFn: (variables: { wordId: string; gloss: string }) =>
+      apiClient.words.updateGloss(variables.wordId, language, variables.gloss),
+    onMutate: ({ wordId }) => {
+      const requestId = Math.floor(Math.random() * 1000000);
+      setGlossRequests((requests) => [...requests, { wordId, requestId }]);
+      return { requestId };
+    },
+    onError: () => {
+      alert('Unknown error occurred.');
+    },
+    onSuccess: (_, { wordId, gloss }, context) => {
+      if (context?.requestId) {
+        setGlossRequests((requests) =>
+          requests.filter((r) => r.requestId !== context.requestId)
+        );
+      }
 
-export default function TranslationView() {
-  const { verse, referenceGlosses } = useLoaderData() as {
-    verse: GetVerseResponseBody;
-    referenceGlosses: GetVerseGlossesResponseBody;
-  };
-  const bookId = parseInt(verse.data.id.slice(0, 2));
-  const hebrew = bookId < 40;
+      const queryKey = ['verse-glosses', language, params.verseId];
+      const query =
+        queryClient.getQueryData<GetVerseGlossesResponseBody>(queryKey);
+      if (query) {
+        const glosses = query.data.slice();
+        const index = glosses.findIndex((g) => g.wordId === wordId);
+        glosses.splice(index, 1, { wordId, gloss });
+        queryClient.setQueryData<GetVerseGlossesResponseBody>(queryKey, {
+          data: glosses,
+        });
+      }
```

Comment:
When the mutation completes, we can use the query key to update the cache for the glosses we are translating.